### ISSUE TITLE
Auto-detect file extension for output file

### ISSUE THREAD

**Issue by nyuszika7h** [streamlink/streamlink#791]
### Checklist

- [ ] This is a bug report.
- [ ] This is a plugin request.
- [x] This is a feature request.
- [x] I used the search function to find already opened/closed issues or pull requests.

### Description

I looked at the documentation and issues but I didn't find a way to automatically choose the correct file extension for the output file when using `-o`. Of course, for example I know that it will be `.ts` for HLS streams, but I might not always know. It would be nice to be able to use `-o filename.{ext}` or something like that. Or an option to remux, since if I use `.mkv` for example, it won't remux to a Matroska container. It will still save fine, but I need to rename the file afterwards for it to make sense.

### Environment details (operating system, python version, etc.)

* Windows 10 x64
* Python 3.6.1
* streamlink 0.5.0

**Comment by bastimeyer** [comment]
IMHO, the video/audio container shouldn't be determined via the extension of the output file (and vice versa). File name extensions are just being used by your OS to link programs with this *potential* file format. It doesn't matter which extension a file has as long as your video player, editing software, etc. is able to correctly parse the container format and video/audio data.

> Or an option to remux

A `--mux` or `--remux` parameter would make more sense where you can choose the container format and properties. Then you can set the file name extension accordingly to your desires.

You can always remux the output on the fly with ffmpeg like this:
```bash
streamlink -O URL QUALITY | ffmpeg -i pipe:0 -c:v copy -c:a copy -f matroska -y file.mkv
```

**Comment by nyuszika7h** [comment]
You're right, but file extensions are still widely used for audio and video files and even if players can handle it, it doesn't make sense to have an extension that doesn't match the container. Thanks for the ffmpeg tip, though.

**Comment by imerr** [comment]
with #1576 merged would it be possible to see similar formatting for the output filename?

(I'm not sure how the internals work or how much work this would be, just me looking for a way to do this and finding the --title flag can do formatting)

**Comment by GrayFace** [comment]
I'll mirror my proposal here then:

**Description**
Weirdly, metadata variables are supported only for rarely used "--title" parameter. They would be very useful for output parameter.

**Proposal:**

Add a new parameter, say "-ox", that would act like "-o", but with variables support.
Expose {title}, {author}, {category} etc. to it.
Also, at some point in the future having upload data/time as a variable(s) would be very useful. I suspect there was no incentive for adding them, because there's not much need for them in "--title" parameter.

**Comment by flower1024** [comment]
i want to second GrayFace. And please add a timestamp variable to.
i would love to generate mkv files from twitch as {year}-{month}-{days}-{hour}:{minute] {author} {game].mkv format

atm my timestamp is always off because of --retry-streams

**Comment by CeruleanSky** [comment]
It seems this would be doable now that `--ffmpeg-fout` is supported but unless I am doing it wrong, when using `-o` the `--ffmpeg-fout` option is ignored.


`streamlink.exe  --ffmpeg-video-transcode copy --ffmpeg-audio-transcode copy --ffmpeg-fout matroska -o -"Test.mkv" "https://www.youtube.com/watch?v=Y1ERSETT0sb" "720p,480p,best"`

and

`streamlink.exe  --ffmpeg-video-transcode copy --ffmpeg-audio-transcode copy --ffmpeg-fout mkv -o -"Test.mkv" "https://www.youtube.com/watch?v=Y1ERSETT0sb" "720p,480p,best"`

and it still returns MPEG TS

```
>file Test.mkv
>Test.mkv: MPEG transport stream data
```

**Comment by bastimeyer** [comment]
> --ffmpeg-fout

This only affects streams which already get muxed by Streamlink, namely instances of [`MuxedStream`](https://github.com/streamlink/streamlink/blob/b9d0d18042fb7326b182d080130a728d99d75587/src/streamlink/stream/ffmpegmux.py#L17) ([`FFMPEGMuxer`](https://github.com/streamlink/streamlink/blob/b9d0d18042fb7326b182d080130a728d99d75587/src/streamlink/stream/ffmpegmux.py#L56)), like DASH streams or HLS streams with additional audio tracks, or streams with additional subtitles. It doesn't make Streamlink remux other streams, like regular HLS streams for example, which are just concatenated MPEG transport stream segments.

The description in the docs is unfortunately not ideal and I didn't notice this when it got merged:
https://streamlink.github.io/cli.html#cmdoption-ffmpeg-fout

The 2.0.0 changelog however describes it properly:
https://streamlink.github.io/changelog.html#streamlink-2-0-0-2020-12-22

**Comment by stobias123** [comment]
I'm seeing the same issues as @CeruleanSky. Even with the updated docs, it's not clear to me when `streamlink` is muxing output. 

Trying to stream from twitch, to MP4 format, but it's not clear to me if this is possible or how I can do this?

**Comment by bastimeyer** [comment]
Let me give an update to this thread, as it's been stale for a while.

----

> Expose {title}, {author}, {category} etc. to it.

1. #3962 has added the `{author}`, `{category}`, `{game}`, `{title}` and `{url}` variables to `--output`/`-o`, `--record`/`-r` and `--record-and-pipe`/`-R`.

2. #3993 has added the `{time}` variable, with optional formatting via `{time:format}` where "format" is a strftime compatible string, eg. `%Y-%m-%d_%H-%M-%S`.

This has already been merged into master and will be available in the next stable release (2.4.0)

----

Now regarding this thread's original topic, namely automatic file name extensions.

After implementing the stuff linked above, I had another look at the file name extensions of the output stream container format yesterday. This is way more difficult to implement than expected, because different stream implementations can output different stream container formats, depending on the source or options being used.

In case of HLS streams for example, MPEG-TS streams and fragmented MPEG-4 streams are both supported, but a stream could also be remuxed with different audio sources or subtitles, resulting in a Matroska container format by default via the FFmpegmuxer, or a completely different one if `--ffmpeg-fout` is set by the user.

Every `Stream` subclass would need to have logic added for determining the name of the format of the input and output stream that could be used by Streamlink CLI for figuring out the file name extension.

There's a problem though: HLS streams for example are simple segmented streams with concatenated segments where it doesn't matter which stream format is being used. This means that the only two options for getting the stream format is either checking for the `Content-Type` HTTP header of the server response or reading the binary header data of the very first segment.

The `Content-Type` header is problematic, because most servers simply send the generic `application/octet-stream` value, which basically means "can be anything, don't care", which is useless. More specific header values could also be incorrect.

The second option of parsing the stream container's header data is also problematic, because of how stream reading and the stream output are implemented in Streamlink CLI. The 8192 bytes of "prebuffer" which Streamlink CLI reads would be enough to figure out the format, but Streamlink CLI needs to decide/prepare/validate the output before it opens/reads the stream data, which means the file name extension in the `FileOutput` can't be chosen before reading the stream data. This would require a re-implementation of how the output works, so that it'll always allocate the prebuffer before initializing the output.

----

As I've already said in my first response to this thread from four years ago,

> File name extensions are just being used by your OS to link programs with this potential file format. It doesn't matter which extension a file has as long as your video player, editing software, etc. is able to correctly parse the container format and video/audio data.

this makes implementing a `{fileext}`/`{ext}` variable pointless in most of the cases. IMO it's not worth the effort of re-writing the current CLI logic and adding a sophisticated system for reading the stream container format headers on top of it when there are lots of problems to overcome like described above.
